File: pt4/1055.shortest-way-to-form-string.java
import java.util.ArrayList;
import java.util.List;

/*
 * 1055. Shortest Way to Form String
 */
class Solution {
    public int shortestWay(String source, String target) {
        @SuppressWarnings("unchecked")
        List<Integer>[] positions = new ArrayList[26];
        for (int i = 0; i < 26; i++) {
            positions[i] = new ArrayList<>();
        }

        for (int i = 0; i < source.length(); i++) {
            positions[source.charAt(i) - 'a'].add(i);
        }

        int subsequences = 1;
        int currentIndex = -1;

        for (int i = 0; i < target.length(); i++) {
            int letter = target.charAt(i) - 'a';
            if (positions[letter].isEmpty()) {
                return -1;
            }

            List<Integer> indexes = positions[letter];
            int next = upperBound(indexes, currentIndex);
            if (next == indexes.size()) {
                subsequences++;
                currentIndex = indexes.get(0);
            } else {
                currentIndex = indexes.get(next);
            }
        }

        return subsequences;
    }

    private int upperBound(List<Integer> values, int target) {
        int left = 0;
        int right = values.size();
        while (left < right) {
            int mid = left + (right - left) / 2;
            if (values.get(mid) <= target) {
                left = mid + 1;
            } else {
                right = mid;
            }
        }
        return left;
    }
}

/*
Interview Explanation

Core idea:
Greedily build each subsequence by taking the earliest possible source index
for each target character. If no occurrence exists after the current index,
start a new subsequence.

Java data structures:
- List<Integer>[] positions stores sorted source indices for each lowercase
  letter.
- Binary search finds the first occurrence greater than currentIndex.

Algorithm:
1. Precompute source positions for every character.
2. Scan target left to right.
3. If a target character does not exist in source, return -1.
4. Use upperBound to find the next usable source occurrence.
5. If none exists, start a new subsequence and use the first occurrence.

Correctness:
Using the earliest possible occurrence of each character leaves the most room
for the rest of the current subsequence, so it can never hurt. A new
subsequence is started only when the current one cannot contain the next target
character. Therefore the greedy process uses the minimum number of
subsequences.

Complexity:
Preprocessing is O(|source|). Each target character performs binary search, so
time is O(|target| log |source|). Space is O(|source|).

Edge cases:
- Missing target character returns -1.
- Target already a subsequence returns 1.
- Repeated wraparounds count multiple subsequences.
*/

File: pt4/1055.shortest-way-to-form-string.go
package main

import "sort"

/*
1055. Shortest Way to Form String
*/
func shortestWay(source string, target string) int {
	positions := make([][]int, 26)
	for i := 0; i < len(source); i++ {
		letter := source[i] - 'a'
		positions[letter] = append(positions[letter], i)
	}

	subsequences := 1
	currentIndex := -1

	for i := 0; i < len(target); i++ {
		letter := target[i] - 'a'
		indexes := positions[letter]
		if len(indexes) == 0 {
			return -1
		}

		next := sort.Search(len(indexes), func(j int) bool {
			return indexes[j] > currentIndex
		})

		if next == len(indexes) {
			subsequences++
			currentIndex = indexes[0]
		} else {
			currentIndex = indexes[next]
		}
	}

	return subsequences
}

/*
Interview Explanation

Core idea:
Build each subsequence greedily by taking the earliest possible source
occurrence for each target character. If the next character cannot appear
after the current source index, start a new subsequence.

Go data structures:
- [][]int positions stores sorted source indices for each lowercase letter.
- sort.Search performs binary search for the first index greater than the
  current source position.

Algorithm:
1. Precompute every source index for each character.
2. Scan target left to right.
3. If a character never appears in source, return -1.
4. Binary search for the next occurrence after currentIndex.
5. If none exists, start a new subsequence and use the first occurrence.

Correctness:
Choosing the earliest possible occurrence leaves the most room for future
characters in the current subsequence, so it can never hurt. A new subsequence
is started exactly when the current one cannot contain the next target
character. Therefore the greedy count is minimal.

Complexity:
Preprocessing is O(len(source)). Each target character uses binary search, so
time is O(len(target) log len(source)). Space is O(len(source)).

Edge cases:
- Missing target character returns -1.
- Target already a subsequence returns 1.
- Repeated wraparounds are counted correctly.
*/
